File: src/camera/camera.cpp
#include "camera.h"

Camera::Camera(glm::vec3 cameraPosition)
{
	this->cameraPosition = cameraPosition;
	this->cameraViewDirection = glm::vec3(0.0f, 0.0f, -1.0f);
	this->cameraUp = glm::vec3(0.0f, 1.0f, 0.0f);
	this->rotationOx = 0.0f;
	this->rotationOy = -90.0f;
	this->right = glm::cross(cameraViewDirection, cameraUp);
}

Camera::Camera()
{
	this ->cameraPosition = glm::vec3(0.0f, 3.0f - 30.0f, 10.0f);
	this ->cameraViewDirection = glm::vec3(0.0f, -1.0f, -1.0f);
	this ->cameraUp = glm::vec3(0.0f, 1.0f, 0.0f);
	this->rotationOx = 0.0f;
	this->rotationOy = -90.0f;
	this->right = glm::cross(cameraViewDirection, cameraUp);
}

Camera::~Camera()
{
}

void Camera::keyboardMoveFront(float cameraSpeed)
{
	cameraPosition += cameraViewDirection * cameraSpeed;
}

void Camera::keyboardMoveBack(float cameraSpeed)
{
	cameraPosition -= cameraViewDirection * cameraSpeed;
}

void Camera::keyboardMoveLeft(float cameraSpeed)
{
	cameraPosition -= glm::normalize(glm::cross(cameraViewDirection, cameraUp)) * cameraSpeed;
}

void Camera::keyboardMoveRight(float cameraSpeed)
{
	cameraPosition += glm::normalize(glm::cross(cameraViewDirection, cameraUp)) * cameraSpeed;
}

void Camera::keyboardMoveUp(float cameraSpeed)
{
	cameraPosition += cameraUp * cameraSpeed;
}

void Camera::keyboardMoveDown(float cameraSpeed)
{
	cameraPosition -= cameraUp * cameraSpeed;
}

void Camera::rotateOx(float angle)
{
	glm::vec3 cameraRight = glm::cross(cameraViewDirection, cameraUp);
	cameraViewDirection = glm::normalize(glm::vec3((glm::rotate(glm::mat4(1.0f), angle, cameraRight) * glm::vec4(cameraViewDirection, 1))));
	cameraUp = glm::normalize(glm::cross(cameraRight, cameraViewDirection));
}

void Camera::rotateOy (float angle)
{
	glm::vec3 cameraRight = glm::cross(cameraViewDirection, cameraUp);
	cameraViewDirection = glm::normalize(glm::vec3((glm::rotate(glm::mat4(1.0f), angle, cameraUp) * glm::vec4(cameraViewDirection, 1))));
	
	cameraRight = glm::normalize(glm::vec3((glm::rotate(glm::mat4(1.0f), angle, cameraUp) * glm::vec4(cameraRight, 1))));
	cameraUp = glm::normalize(glm::cross(cameraRight, cameraViewDirection));
}

void Camera::mouseMovement(float xpos, float ypos, int windowWidth, int windowHeight)
{
	float sensitivity = 0.1f;

	double xoffset = -(xpos - windowWidth / 2) * sensitivity;
	double yoffset = (windowHeight / 2 - ypos) * sensitivity;

	rotationOy += xoffset;
	rotationOx -= yoffset;

	update(xoffset, yoffset);
}

glm::mat4 Camera::getViewMatrix()
{
	return glm::lookAt(cameraPosition, cameraPosition + cameraViewDirection, cameraUp);
}

glm::vec3 Camera::getCameraPosition()
{
	return cameraPosition;
}

glm::vec3 Camera::getCameraViewDirection()
{
	return cameraViewDirection;
}

glm::vec3 Camera::getCameraUp()
{
	return cameraUp;
}

void Camera::update(float xoffset, float yoffset)
{
	glm::vec3 newViewDirection;
	newViewDirection.x = cos(glm::radians(rotationOy)) * sin(glm::radians(rotationOx));
	newViewDirection.y = sin(glm::radians(rotationOy));
	newViewDirection.z = cos(glm::radians(rotationOy)) * cos(glm::radians(rotationOx));
	
	cameraViewDirection = glm::mat3(glm::rotate(yoffset, right)) * glm::mat3(glm::rotate(xoffset, cameraUp)) * cameraViewDirection;
	right = glm::normalize(glm::cross(cameraViewDirection, cameraUp));
}

File: Graphics/Camera.cs
using System.Numerics;

public class Camera
{
    public Vector3 Position { get; set; }
    public Vector3 ViewDirection { get; set; }
    public Vector3 Up { get; set; }
    public Vector3 Right { get; set; }

    public float RotationOx { get; set; }
    public float RotationOy { get; set; }

    public Camera()
    {
        Position = new Vector3(0.0f, 0.0f, -65.0f);
        ViewDirection = new Vector3(0.0f, 0.0f, -1.0f);
        Up = new Vector3(0.0f, 1.0f, 0.0f);
        RotationOx = -90.0f;
        RotationOy = 0.0f;
        Right = Vector3.Cross(ViewDirection, Up);
    }

    public Camera(Vector3 cameraPosition)
    {
        Position = cameraPosition;
        ViewDirection = new Vector3(0.0f, 0.0f, -1.0f);
        Up = new Vector3(0.0f, 1.0f, 0.0f);
        RotationOx = 0.0f;
        RotationOy = -90.0f;
        Right = Vector3.Cross(ViewDirection, Up);
    }

    public void KeyboardMoveFront(float cameraSpeed)
    {
        Position += ViewDirection * cameraSpeed;
    }

    public void KeyboardMoveBack(float cameraSpeed)
    {
        Position -= ViewDirection * cameraSpeed;
    }

    public void KeyboardMoveLeft(float cameraSpeed)
    {
        Position -= Vector3.Normalize(Vector3.Cross(ViewDirection, Up)) * cameraSpeed;
    }

    public void KeyboardMoveRight(float cameraSpeed)
    {
        Position += Vector3.Normalize(Vector3.Cross(ViewDirection, Up)) * cameraSpeed;
    }

    public void KeyboardMoveUp(float cameraSpeed)
    {
        Position += Up * cameraSpeed;
    }

    public void KeyboardMoveDown(float cameraSpeed)
    {
        Position -= Up * cameraSpeed;
    }

    /*
    public void RotateOx(float angle)
    {
        Vector3 cameraRight = Vector3.Cross(ViewDirection, Up);
        ViewDirection = Vector3.Normalize(new Vector3(Matrix3.CreateFromAxisAngle(cameraRight, angle) * new Vector4(ViewDirection, 1)));
        Up = Vector3.Normalize(Vector3.Cross(cameraRight, ViewDirection));
    }

    public void RotateOy(float angle)
    {
        Vector3 cameraRight = Vector3.Cross(ViewDirection, Up);
        ViewDirection = Vector3.Normalize(new Vector3(Matrix4.CreateFromAxisAngle(Up, angle) * new Vector4(ViewDirection, 1)));

        cameraRight = Vector3.Normalize(new Vector3(Matrix4x4.CreateFromAxisAngle(Up, angle) * new Vector4(cameraRight, 1)));
        Up = Vector3.Normalize(Vector3.Cross(cameraRight, ViewDirection));
    }

    public void MouseMovement(float xpos, float ypos, int windowWidth, int windowHeight)
    {
        float sensitivity = 0.01f;

        float xoffset = -(xpos - windowWidth / 2) * sensitivity;
        float yoffset = (windowHeight / 2 - ypos) * sensitivity;

        RotationOy += xoffset;
        RotationOx -= yoffset;

        Update(xoffset, yoffset);
    }

    public void Update(float xoffset, float yoffset)
    {
        Vector3 newViewDirection;
        newViewDirection.X = (float)Math.Cos(RotationOy) * (float)Math.Sin(RotationOx);
        newViewDirection.Y = (float)Math.Sin(RotationOy);
        newViewDirection.Z = (float)Math.Cos(RotationOy) * (float)Math.Cos(RotationOx);

        ViewDirection = Matrix3x3.CreateRotationY(yoffset) * Matrix3x3.CreateRotationX(xoffset) * ViewDirection;
        Right = Vector3.Normalize(Vector3.Cross(ViewDirection, Up));
    }
    */
}